PD 1: Replace a failed drive (#420)
portworx/pxdocs: content/reference/troubleshoot/replace-a-failed-drive/_index.md

---
title: Remove or replace failed drive
description: Deal with a failed drive by either removing it or replacing it
keywords: Portworx, troubleshoot
series: troubleshoot-portworx
---

> When a drive fails, Portworx continues to operate using available replicas on other nodes. Pods associated with the failed drive may crash or be otherwise impacted. To fully recover from a drive failure, you must replace or remove the failed drive from Portworx, and how you recover depends on the pool which contains the failed drive.

Reviewer: ```suggestion
When a drive fails, Portworx continues to operate using available replicas on other nodes. To fully recover from a drive failure, you must replace or remove the failed drive from Portworx, and how you recover depends on the pool which contains the failed drive.
```

Follow the steps in one of the sections below to either:

* Replace a drive that belongs to a pool which hosts Portworx metadata
* Replace a drive that belongs to a non-metadata pool

## Replace a drive that belongs to a pool which hosts Portworx metadata

If the drive belongs to a pool that hosts Portworx metadata, then you must:

* Remove the node from the cluster
* Replace the failed drive
 
### Prerequisites

Before you can replace the drive, you must meet the following prerequisites:

* Ensure that the pool hosts metadata by entering the following command:

    ```text
    pxctl service pool show
    ```

    ```output
    PX drive configuration:
    Pool ID: 1
    UUID: 86d5d105-2eff-4dfb-b842-eca86906c921
    IO Priority: LOW
    Labels: iopriority=LOW,medium=STORAGE_MEDIUM_MAGNETIC
    Size: 20 GiB
    Status: Online
    Has metadata: Yes
    Drives:
    1: /dev/sdc, 8.0 GiB allocated of 20 GiB, Online
    Cache Drives:
    ```

    In the output above, note the line that reads "Has metadata: Yes".

>     <!-- Is this output okay? -->

Author: @prabirpaul Please see my comment.

  * Find the volumes that match the pool identifier
  * Remove all replicas of the volumes in this pool
  * Delete the pool
  * (Optional step) Replace the drive
  * Re-add the drives

### Prerequisites

Before you can replace the drive, you must meet the following prerequisites:

* The failed drive must belong to a node that contains more than one storage pool. You cannot transition to a storage less node using this procedure.
* Ensure that the pool doesn't contain metadata by entering the following command:

    ```text
    pxctl service pool show
    ```

    ```output
    PX drive configuration:
    Pool ID: 1
    UUID: 86d5d105-2eff-4dfb-b842-eca86906c921
    IO Priority: LOW
    Labels: iopriority=LOW,medium=STORAGE_MEDIUM_MAGNETIC
    Size: 20 GiB
    Status: Online
    Has metadata: No
    Drives:
    1: /dev/sdc, 8.0 GiB allocated of 20 GiB, Online
    Cache Drives:
    ```

    In the output above, note the line that reads "Has metadata: No".

### Replace the failed drive

Perform the following steps to replace the failed drive:

1. Enter the following `pxctl volume list` command, specifying your node identifier, to list the volumes in this pool:
  ```text

>   pxctl volume list --node_id <node-id>

Reviewer: ```suggestion
  pxctl volume list --node-id <node-id>
```

  ```

>   <!-- Need output. Running this on our test cluster returns an empty list. -->

Author: @prabirpaul @nathan-px Please see my comment.

Reviewer: ```suggestion
pxctl volume list --node-id c95bb1d2-4803-450b-9117-d32e4aa08dae
IO_PRIORITY	STATUS					SNAP-ENABLED	
50487189447950970	pvc-6535e2d1-3c9c-11ea-9a49-525400d7f2ff	1 GiB	2	no	no		LOW		up - attached on 192.168.121.139	no
```

> 2. Look for the pool option. If none exists, then you must parse through the output to find the volumes that match the pool identifier. <!-- Depending on how the output looks like, this could be reworded.-->

Author: @prabirpaul @nathan-px Please see my comment.

Reviewer: I couldn't find this either. There doesn't appear to be any output of associated pools from the volume json

Reviewer 2: from pxctl -j v i :
    "ReplicaSetCreateMid": "a4b0f686-861e-4f4c-b978-b31973bead90,b78551cc-e48d-424b-88b2-4a69c6f0eb4d",
    "ReplNodePools": "0,1",
=> Pool 0 on a4b0f686-861e-4f4c-b978-b31973bead90
=> pool 1 on b78551cc-e48d-424b-88b2-4a69c6f0eb4d

3. Manually remove all replicas of the volumes in this pool. Refer to the [Decreasing the replication factor](/reference/cli/updating-volumes/#decreasing-the-replication-factor) section for details.

4. Delete the pool by entering the `pxctl service pool delete` command with the identifier of your pool as an argument. The following example deletes the pool that has the pool identifier of one (`Pool ID: 1`):
    ```text
    pxctl service pool delete 1
    ```

5. Replace the failed drive ensuring that it's the same capacity and type as the failed one. Note that you can also reform the pool without the failed drive, resulting in a **functional pool with one less drive**.

6. You can re-add the drives by entering the `pxctl service drive add` command, specifying the `--drive` option with the paths to your drives. The following command adds three drives:
  ```text
  pxctl service drive add --drive /dev/sdc /dev/sdd /dev/sde
  ```

>   <!-- Is this ^^ correct? -->

Author: @prabirpaul @nathan-px Please see my comment

Reviewer: looks like it to me

2. Determine if the pool containing that drive hosts metadata by entering the `pxctl service pool show` command and looking for the `Has metadata` field in the output:

    ```text
    pxctl service pool show
    ```
    ```output
    PX drive configuration:
    Pool ID: 1
    UUID: 86d5d105-2eff-4dfb-b842-eca86906c921
    IO Priority: LOW
    Labels: iopriority=LOW,medium=STORAGE_MEDIUM_MAGNETIC
    Size: 20 GiB
    Status: Online
    Has metadata: Yes
    Drives:
    1: /dev/sdc, 8.0 GiB allocated of 20 GiB, Online
    Cache Drives:
    ```

    In the output above, note the line that reads "Has metadata: Yes".

If your service pool has metadata, follow the procedure to **Replace a drive that belongs to a pool which hosts Portworx metadata**. If your service pool does not have metadata, follow the procedure to **Replace a drive that belongs to a non-metadata pool**.

## Replace a drive that belongs to a pool which hosts Portworx metadata

If the drive belongs to a pool that hosts Portworx metadata, then you must remove the node from the cluster and replace the failed drive.

Perform the following steps to replace the failed drive:

> 1. Use the Use [Node decommission workflow](/portworx-install-with-kubernetes/operate-and-maintain-on-kubernetes/uninstall/decommission-a-node/#step-1-migrate-application-pods-using-portworx-volumes-that-are-running-on-this-node) to remove a node from the cluster. In this workflow, you reduce the HA level for all replicated volumes which reside on the node and restore the volumes if enough storage nodes are available in the cluster.

Reviewer 2: The HA reduce and increase happens automatically, no user intervention is required.

> 1. Use the Use [Node decommission workflow](/portworx-install-with-kubernetes/operate-and-maintain-on-kubernetes/uninstall/decommission-a-node/#step-1-migrate-application-pods-using-portworx-volumes-that-are-running-on-this-node) to remove a node from the cluster. In this workflow, you reduce the HA level for all replicated volumes which reside on the node and restore the volumes if enough storage nodes are available in the cluster.

Reviewer: ```suggestion
1. Use the Use [Node decommission workflow](/portworx-install-with-kubernetes/operate-and-maintain-on-kubernetes/uninstall/decommission-a-node/#step-1-migrate-application-pods-using-portworx-volumes-that-are-running-on-this-node) to remove a node from the cluster. This reduces the HA level for all replicated volumes which reside on the node and restores the volumes if enough storage nodes are available in the cluster.
```

> 2. Replace the failed drive and add the node back into the cluster. Note that Portworx runs as a `DaemonSet` in Kubernetes, so when you add a node or a worker to your Kubernetes cluster, you don't need to explicitly run Portworx on it.

Reviewer 2: point #5 below applies here as well.

## Replace a drive that belongs to a non-metadata pool

If the drive belongs to a non-metadata pool, then you must delete the affected storage pool.

{{<info>}}
**NOTE:** The failed drive must belong to a node that contains more than one storage pool. You cannot transition to a storageless node using this procedure.
{{</info>}}

Perform the following steps to replace the failed drive:

> 1. Enter the following `pxctl volume list` command, specifying your own node identifier to list the volumes in this pool:

Reviewer: here we're listing all of the volumes on this node
```suggestion
1. Enter the following `pxctl volume list` command, specifying your own node identifier to list the volumes in this node:
```

  ```text
  pxctl volume list --node-id <node-id>
  ```
  ```output
  IO_PRIORITY	STATUS					SNAP-ENABLED
  50487189447950970	pvc-6535e2d1-3c9c-11ea-9a49-525400d7f2ff	1 GiB	2	no	no		LOW		up - attached on 192.168.1.1	no
  ```

> 2. Look for the pool option. If none exists, then you must parse through the output to find the volumes that match the pool identifier. <!-- Depending on how the output looks like, this could be reworded.-->

Reviewer: Of the list of volumes in the node, list the volumes which reside on the target pool (the pool with the failing drive)

> 

Reviewer: ```suggestion
```text
pxctl -j volume inspect
```
```output
"ReplicaSetCreateMid": "a4b0f686-861e-4f4c-b978-b31973bead90,b78551cc-e48d-424b-88b2-4a69c6f0eb4d",
"ReplNodePools": "0,1",
```
```

Reviewer 2: The above 2 steps can be done like this:
- Get the pool uid of the target pool
[root@ip-70-0-52-144 ~]# pxctl sv pool show
PX drive configuration:
Pool ID: 0 
	**UUID:  63e528b8-bc2e-484e-a01e-91e5108ebba5** 
	IO Priority:  HIGH 
	Labels:  iopriority=HIGH,medium=STORAGE_MEDIUM_SSD 
	Size: 384 GiB 
	Status: Online 
	Has metadata:  Yes 
	Drives:
	1: /dev/sdc, 7.0 GiB allocated of 128 GiB, Online
	2: /dev/sdd, 6.0 GiB allocated of 128 GiB, Online
	3: /dev/sde, 7.0 GiB allocated of 128 GiB, Online
	Cache Drives:
	No Cache drives found in this pool
- Use the uuid from above to list the volumes which are on the pool
[root@ip-70-0-52-144 ~]# pxctl v l --pool-uid 63e528b8-bc2e-484e-a01e-91e5108ebba5
